src/components/viewport.tsx

```diff
@@ -181,6 +182,31 @@ function Viewport(props: ViewportProps) {
     ],
   )
 
+  const getBottomShadowOfCurrentPiece = useCallback(() => {
+    for (let i = currentPieceInViewport.y + 1; i <= VIEWPORT_HEIGHT; i++) {
+      const futureCurrentPieceInViewport = {
+        ...currentPieceInViewport,
+        y: i,
+      }
+
+      const futurePieceHitsOtherPieces =
+        checkIfPieceHitsOtherPieces(
+          { ...futureCurrentPieceInViewport, y: futureCurrentPieceInViewport.y + 1 },
+          currentGameState,
+        ) || checkIfPieceHitsBottom(futureCurrentPieceInViewport)
+      if (futurePieceHitsOtherPieces) {
+        setShadowPiece({
+          ...futureCurrentPieceInViewport,
+          piece: {
+            ...futureCurrentPieceInViewport.piece,
+            color: "#000",
+          },
+        })
+        break
+      }
+    }
+  }, [currentGameState, currentPieceInViewport])
```

Comment:
This should be done with `useMemo` since the ghost piece always derives from currentPiece.